lightning-net/src/lib.rs

```diff
@@ -0,0 +1,1144 @@
+impl PartialEq for SyncSocketDescriptor {
+    fn eq(&self, other: &Self) -> bool {
+        self.id == other.id
+    }
+}
+impl Eq for SyncSocketDescriptor {}
+impl hash::Hash for SyncSocketDescriptor {
+    fn hash<H: hash::Hasher>(&self, state: &mut H) {
+        self.id.hash(state)
+    }
+}
+impl SyncSocketDescriptor {
+    fn new(
+        connection_id: u64,
+        tcp_disconnector: TcpDisconnectooor,
+        reader_cmd_tx: Sender<ReaderCommand>,
+        writer_cmd_tx: Sender<WriterCommand>,
+        write_data_tx: Sender<Vec<u8>>,
+    ) -> Self {
+        Self {
+            id: connection_id,
+            tcp_disconnector,
+            reader_cmd_tx,
+            writer_cmd_tx,
+            write_data_tx,
+        }
+    }
+}
+impl SocketDescriptor for SyncSocketDescriptor {
+    /// Attempts to queue up some data from the given slice for the `Writer` to
+    /// send. Returns the number of bytes that were successfully pushed to the
+    /// `write_data` channel.
+    ///
+    /// This implementation never calls back into the PeerManager directly,
+    /// thereby preventing reentrancy / deadlock issues. Instead, any commands
+    /// to be processed and data to be sent are dispatched to the Reader or
+    /// Writer via crossbeam channels.
+    ///
+    /// Additionally, sending across the crossbeam channels is done exclusively
+    /// with non-blocking try_send()s rather than blocking send()s, to ensure
+    /// that this function always returns immediately, thereby also reducing the
+    /// amount of time that the PeerManager's internal locks are held.
+    fn send_data(&mut self, data: &[u8], resume_read: bool) -> usize {
+        if resume_read {
+            let _ = self.reader_cmd_tx.try_send(ReaderCommand::ResumeRead);
+        }
+
+        if data.is_empty() {
+            return 0;
+        }
+
+        if self.write_data_tx.is_empty() {
+            // The data must be copied into the channel since a &[u8] reference
+            // cannot be sent across threads. This incurs a small amount of overhead.
+            match self.write_data_tx.try_send(data.to_vec()) {
+                Ok(()) => data.len(),
+                Err(e) => match e {
+                    TrySendError::Full(_) => {
+                        // This could only happen if another Sender pushed into
+                        // the channel in between the if check above and now - a
+                        // TOCTTOU error. This really shouldn't happen, but
+                        // let's just proceed normally: pause reads and return 0
+                        let _ = self.reader_cmd_tx.try_send(ReaderCommand::PauseRead);
+                        0
+                    }
+                    TrySendError::Disconnected(_) => {
+                        // This might happen if the Writer detected a disconnect and
+                        // shut down on its own. Return 0.
+                        0
+                    }
+                },
+            }
```

Comment:
since we're only sending data on the channel and the channel is `bounded(1)`, we don't need to check for `is_empty()` first. the flow is a bit nicer then:

```rust
match write_tx.try_send(data) {
  Ok(()) => data.len(),
  Err(TrySendError::Full(_)) => self.pause_read()
  Err(TrySendError::Disconnected(_)) => 0,
}
```

Reply:
Nice 👌

I originally wanted to remove the unneeded `clone()` in the case that the channel was already full, but this is much nicer and avoids the weird tocttou comment.